### DISCUSSION TITLE
One input stream to multiple output, output will affect each other.

### DISCUSSION THREAD

**Discussion by liguangcheng** [fluent/fluent-bit#7551] [Q&A]
I configured one input and stream to 2 http outputs, when I down one of http output, after running a period of time, the normal http output will also stop, can anyone face this issue? how do I configure the fluent bit.

following are my configuration.
```
[INPUT]
    Name                tail
    alias               egress
    Path                /opt/fluent-bit/bin/egress/*/*/*
    Key                 event
    DB                  /opt/fluent-bit/bin/egress/fluentbit-egress.db
    Tag                 EGRESS
    Mem_Buf_Limit       20M
    Skip_Long_Lines     on
    read_from_head      true
    Exclude_Path        *.gz,*.zip

[FILTER]
    Name                    modify
    match                   EGRESS
    set     host            hostname
    set     sourcetype      egress

[OUTPUT]
    name                http
    match               EGRESS
    alias               splunk
    json_date_key       time
    host                127.0.0.1
    port                1338
    uri                 /long
    header              Authorization Splunk hehehe
    format              json
    tls                 Off
    tls.verify          False
    Retry_Limit         False

[OUTPUT]
    name                http
    match               EGRESS
    alias               kafka
    json_date_key       time
    host                127.0.0.1
    port                1339
    uri                 /long
    header              Authorization Splunk abcd1234
    format              json
    tls                 Off
    tls.verify          False
    Retry_Limit         False
```

**Comment by patrick-stephens** [comment]
You probably want to review https://docs.fluentbit.io/manual/administration/backpressure

Also note that you're configuring infinite retries not disabling them: https://docs.fluentbit.io/manual/administration/scheduling-and-retries#configuring-retries - you want `no_retries` for that.

**Reply by liguangcheng** [reply]
Thank you for your suggestion. In the production environment, data cannot be lost, even if the backend is down for a long time, so I will set retry to infinite.

**Reply by patrick-stephens** [reply]
Sure, but then you hit a buffer limit you have speciified.

You need to decide on what to do: if you never want data lost then you need to ensure your buffers can cope with however long the downtime is.

The chunks are waiting to be sent as your output is down. Tail is configured only to detain a certain amount of chunks in the buffer so when it reaches that limit it pauses - it will restart from where it paused (as long as the file is still there) once chunks start being sent.

**Reply by liguangcheng** [reply]
Thank you for your quick response, as you said the root cause is that the retry is infinite which lead buffer full, when i set the retry to specific 3 or 2 times, one of normal output works well.

Also, maybe to bother you, I have another problem.
I set an input configuration to ingest a file which all lines are Json object string, when I start the fluent bit to read file to output to a http server, I found the contents of the file could not be completely sent, I check the metric `fluentbit_input_records_total` the count smaller than the file line count, even if I stop the test the file size not the grow up the metrics count still not same with the file line count

Can you help me analyze what causes it? thanks a lot.
To add, when I copy the content of a file to another file, the new file content can be delivery complete, I wonder if it has something to do with reading while writing.

**Reply by patrick-stephens** [reply]
It should not but depends what you mean. Fluent Bit reads line by line so if you append should be fine but if you are destroying the contents of the file whilst it is reading it then yeah lots of stuff will go wrong with any tool.

**Reply by liguangcheng** [reply]
Thank you for your response. :) I did not destroy the contents of the file.

Regarding log missing issue, I think I found the root cause that is after the high-speed writing file stops, fluent cannot send the content at the back of the file, I use fluent bit of version 2.0.8, for this version, fluent lack of a mechanism that ingest file that not append new line but still has lines that not sent.

I check the code fount that half year ago `flb_tail_fs_inotify_init` function introduce a background operation

```
ret = flb_input_set_collector_time(in, in_tail_progress_check_callback,
                                       ctx->progress_check_interval,
                                       ctx->progress_check_interval_nsec,
                                       config);
```
this timer will trigger the `in_tail_progress_check_callback` function per 2 seconds to ingest the line of file that has long time no touch

1. version 2.0.8 has no such compensation
2. `in_tail_progress_check_callback` callback running cycle is too slow(2s), low performance

**Reply by liguangcheng** [reply]
Event if introduce function `in_tail_progress_check_callback ` to remedies, fluent bit still faces the log may missing due to high speed write while low speed stream to output once write stop after a while.

**Reply by patrick-stephens** [reply]
I would use the latest version, nothing will be updated for 2.0.8 only new versions.

**Reply by liguangcheng** [reply]
Yes, I will upgrade the fluent bit later, but I am worried that the latest version will not be able to solve this problem from the root.

**Reply by liguangcheng** [reply]
@patrick-stephens  sorry to disturb you again, following is part of my configuration, I define two pipelines for same path, that means two independent INPUT to monitor same log directory, and according to different tag the same log will be route to different backend system like http and Kafka, this works but I face a serious problem that is consume huge memory, while only one input just only consume little memory, can you please help to analysis, than you very much

**the reason why I want to define two pipelines for one directory is in the case of only one input, if one of the two backends stops, the other will also be affected.**

```
[INPUT]
    name                tail
    tag                 event.kafka.ingress
    alias               kafka.ingress
    buffer_chunk_size   32k
    buffer_max_size     256k
    read_from_head      true
    refresh_interval    5
    rotate_wait         10
    skip_empty_lines    off
    skip_long_lines     true
    key                 message
    db                  /var/log/logshipper/kafka.ingress.db
    db.sync             normal
    db.locking          true
    db.journal_mode     off
    path                /var/log/ses/ingress/*/*/*/*,/var/log/ses/ingress/*/*/*/*/*,/var/log/ses/ingress/*/*/*/*/*/*
    exclude_path        /var/log/logshipper/logshipper.log,/var/log/ses/ingress/*.gz,/var/log/ses/ingress/*.tgz
    mem_buf_limit       20MB
    parser              json
    ignore_older        11m

[INPUT]
    name                tail
    tag                 event.http.ingress
    alias               http.ingress
    buffer_chunk_size   32k
    buffer_max_size     256k
    read_from_head      true
    refresh_interval    5
    rotate_wait         10
    skip_empty_lines    off
    skip_long_lines     true
    key                 message
    db                  /var/log/logshipper/ingress.db
    db.sync             normal
    db.locking          true
    db.journal_mode     off
    path                /var/log/ses/ingress/*/*/*/*,/var/log/ses/ingress/*/*/*/*/*,/var/log/ses/ingress/*/*/*/*/*/*
    path_key            fluentbitLogFilePath
    exclude_path        /var/log/logshipper/logshipper.log,/var/log/ses/ingress/*.gz,/var/log/ses/ingress/*.tgz
    mem_buf_limit       20MB
    parser              json
    ignore_older        11m

[FILTER]
    name                nest
    match               event.kafka.ingress
    operation           nest
    wildcard            *
    nest_under          message
[FILTER]
    name                modify
    match               event.kafka.ingress
    set host            ${POD_NAME}
    set messageType     ingress

[FILTER]
    name                nest
    match               event.http.ingress
    operation           nest
    wildcard            *
    nest_under          message
[FILTER]
    name                modify
    match               event.http.ingress
    set msgtype         ingress
    set host            ${POD_NAME}
    set index           dfsIndex
[FILTER]
    name                nest
    match               event.http.ingress
    operation           nest
    wildcard            message
    wildcard            msgtype
    nest_under          event

[OUTPUT]
    name                kafka
    match               event.kafka.ingress
    alias               kafka.ingress
    format              json
    brokers             lgc-kafka.stc-n41-lsv4.svc.cluster.local:9092
    topics              lgc-events
    timestamp_key       time
    rdkafka.compression.codec                        gzip
    rdkafka.client.id                                ${POD_NAME}
    rdkafka.debug                                    broker,topic,consumer
    rdkafka.message.max.bytes                        1280000
    rdkafka.request.required.acks                    1
    rdkafka.queue.buffering.max.messages             10000

[OUTPUT]
    name                http
    match               event.http.ingress
    alias               http.ingress
    json_date_key       time
    json_date_format    epoch
    host                lsv-test-helm-eric-ses-backend-simulator.stc-n41-lsv
    uri                 /services/collector/event
    port                2222
    retry_limit         false
    log_response_payload false
    header              Authorization Splunk abcd1234
    format              json
    tls                 off
    tls.verify          false

```

**Reply by liguangcheng** [reply]
To add, actually in my environment, I have ingress, egress, tel, event 4 type log, so that means when I fully configured then there are 8 pipelines in fluent bit, the above configuration just a example that show ingress how to configure.

When I just configure 4 pipelines for example, just route logs to http backend system, the memory consumption is relative small(500M/process), but when I fully configure 8 pipelines then the memory consumption will really high(7.5G/process).
